allennlp/data/data_loaders/pytorch_data_loader.py

```diff
@@ -0,0 +1,224 @@
+from typing import List, Iterator
+
+from torch.utils import data
+
+from allennlp.common.lazy import Lazy
+from allennlp.common.tqdm import Tqdm
+from allennlp.data.instance import Instance
+from allennlp.data.dataset_readers import DatasetReader, WorkerInfo
+from allennlp.data.samplers import PyTorchSampler, PyTorchBatchSampler
+from allennlp.data.vocabulary import Vocabulary
+from allennlp.data.data_loaders.data_loader import DataLoader, allennlp_collate
+
+
+class AllennlpDataset(data.Dataset):
+    def __init__(self, instances: List[Instance], vocab: Vocabulary = None):
+        self.instances = instances
+        self.vocab = vocab
+
+    def __getitem__(self, idx) -> Instance:
+        if self.vocab is not None:
+            self.instances[idx].index_fields(self.vocab)
+        return self.instances[idx]
+
+    def __len__(self):
+        return len(self.instances)
+
+    def __iter__(self) -> Iterator[Instance]:
+        """
+        Even though it's not necessary to implement this because Python can infer
+        this method from `__len__` and `__getitem__`, this helps with type-checking
+        since `AllennlpDataset` can be considered an `Iterable[Instance]`.
+        """
+        yield from self.instances
+
+    def index_with(self, vocab: Vocabulary):
+        self.vocab = vocab
+
+
+class AllennlpLazyDataset(data.IterableDataset):
+    """
+    # Parameters
+
+    instance_generator : `Callable[[str], Iterable[Instance]]`
+        A factory function that creates an iterable of `Instance`s from a file path.
+        This is usually just `DatasetReader._instance_iterator`.
+    file_path : `str`
+        The path to pass to the `instance_generator` function.
+    vocab : `Vocab`, optional (default = `None`)
+        An optional vocab. This can also be set later with the `.index_with` method.
+    """
+
+    def __init__(self, reader: DatasetReader, file_path: str, vocab: Vocabulary = None) -> None:
+        super().__init__()
+        self.reader = reader
+        self.file_path = file_path
+        self.vocab = vocab
+
+    def __iter__(self) -> Iterator[Instance]:
+        for instance in Tqdm.tqdm(self.reader.read(self.file_path), desc="reading instances"):
+            if self.vocab is not None:
+                instance.index_fields(self.vocab)
+            yield instance
+
+    def index_with(self, vocab: Vocabulary):
+        self.vocab = vocab
+
+
+def allennlp_worker_init_fn(worker_id):
+    worker_info = data.get_worker_info()
+    dataset = worker_info.dataset
+    if isinstance(dataset, AllennlpLazyDataset):
+        dataset.reader._set_worker_info(
+            WorkerInfo(num_workers=worker_info.num_workers, worker_id=worker_id)
+        )
+
+
+@DataLoader.register("pytorch_data_loader", constructor="from_partial_objects")
```

Comment:
We generally don't put the type into the registrable name. I think this should just be called `"pytorch"` then.